core/arch/mangle_shared.c

```diff
@@ -771,6 +771,122 @@ mangle_syscall_code(dcontext_t *dcontext, fragment_t *f, byte *pc, bool skip)
+            REPORT_FATAL_ERROR_AND_EXIT(
+                RSEQ_BEHAVIOR_UNSUPPORTED, 3, get_application_name(),
+                get_application_pid(),
+                "Malformed rseq endpoint: not on instruction boundary");
+            ASSERT_NOT_REACHED();
+        }
+#    ifdef X86
+        /* We just ran the instrumented version of the rseq code, with the stores
+         * removed.  Now we need to invoke it again natively for real.  We have to
+         * invoke the abort handler, not the rseq start, as it may perform some setup.
+         *
+         * XXX i#2350: We assume the code can handle "aborting" once every time and
+         * still make forward progress.  We document this in our Limitations section.
+         * We also ignore the flags for when to restart.  It's possible the app
+         * disabled restarts on preempts and migrations and can't handle our restart
+         * here.
+         */
+        LOG(THREAD, LOG_INTERP, 4, "mangle: inserting call to native rseq " PFX "\n",
+            handler);
+        RSTATS_INC(num_rseq_native_calls_inserted);
+        /* For simplicity in this first version of the code, we assume call-return
+         * semantics for the abort handler + rseq region.  We create an extra frame
+         * and assume that causes no problems.  We assume the native invocation will
+         * come back to us.
+         * TODO i#2350: Make a copy of the rseq code and append a "return" as a jmp*
+         * through TLS.  Make a copy of the abort handler too.  Decode it and ensure
+         * it transitions directly to the rseq code (if not, bail).  "Call" the
+         * copied abort handler+rseq by setting the retaddr in TLS and jumping.  Then
+         * we have a guaranteed return and no extra frame, with verified assumptions.
+         */
+        instr_t check;
+        instr_init(dcontext, &check);
+        if (decode_cti(dcontext, end, &check) == NULL || !instr_is_return(&check)) {
+            REPORT_FATAL_ERROR_AND_EXIT(RSEQ_BEHAVIOR_UNSUPPORTED, 3,
+                                        get_application_name(), get_application_pid(),
+                                        "Rseq sequences must end with a return");
+            ASSERT_NOT_REACHED();
+        }
+        instr_free(dcontext, &check);
+        /* We assume that by making this a block end, clients will restore app state
+         * before this native invocation.
+         * TODO i#2350: Take some further action to better guarantee this in the face
+         * of future drreg optimizations, etc.  Do we need new interface features, or
+         * do we live with a fake app jump or sthg?
+         */
+        /* A direct call may not reach so we clobber a scratch register. */
+        insert_mov_immed_ptrsz(dcontext, (ptr_int_t)handler, opnd_create_reg(DR_REG_RAX),
+                               ilist, next_instr, NULL, NULL);
+        /* Set up the frame and stack alignment.  We assume the rseq code was a leaf
```

Comment:
A bitmask requires saving the prior rsp and restoring it.  We could stick it in a spill slot.  However, I expected this inserted call to be temporary, because I was hoping to come up with a scheme that does not use an extra call frame at all -- though I don't have a great one yet since making a copy requires changing the rseq data struct, which requires locating it; the alternative of patching the app code is not super appealing either.  For now I will put in a comment at least about guaranteeing alignment if we stick with a call longer-term.

Reply:
In PR #3757.